### PR TITLE
[Xcodeproj] Don't add excluded directories as blue folders

### PR DESCRIPTION
**Pull request by emlai**
Xcode project generation now considers the [`exclude`](https://github.com/apple/swift-package-manager/blob/master/Documentation/Reference.md#exclude) list when deciding which extra directories to add as folder references.

### CODE REVIEW

File: Tests/XcodeprojTests/GenerateXcodeprojTests.swift [hunk 1]
```diff
@@ -44,6 +44,26 @@ class GenerateXcodeprojTests: XCTestCase {
...
+            mktmpdir { dstdir in
+                try makeDirectories(dstdir.appending(component: "excludeMe"))
+
+                func dummy() throws -> [Module] {
+                    return [try SwiftModule(name: "DummyModuleName", sources: Sources(paths: [], root: dstdir))]
+                }
+
+                let projectName = "DummyProjectName"
+                let dummyPackage = Package(manifest: Manifest(path: dstdir, url: dstdir.asString, package: PackageDescription.Package(name: "Foo", exclude: ["excludeMe"]), products: [], version: nil), path: dstdir, modules: [], testModules: [], products: [])
+                let graph = PackageGraph(rootPackage: dummyPackage, modules: try dummy(), externalModules: [])
```
Comment (aciidb0mb3r) [inline Tests/XcodeprojTests/GenerateXcodeprojTests.swift]: Above code can be simplified to:
```swift
let fileSystem = InMemoryFileSystem(emptyFiles: "/source.swift", "/exculdeMe/anchor.txt")
let graph = try loadMockPackageGraph(["/": name: "Foo", exclude: ["excludeMe"]], root: "/", in: fileSystem)
```

File: Tests/XcodeprojTests/GenerateXcodeprojTests.swift [hunk 2]
```diff
@@ -44,6 +44,26 @@ class GenerateXcodeprojTests: XCTestCase {
...
+                let outpath = try Xcodeproj.generate(outputDir: dstdir, projectName: projectName, graph: graph, options: XcodeprojOptions())
+
+                let pbxproj = try String(contentsOfFile: outpath.appending(component: "project.pbxproj").asString)
+                XCTAssert(!pbxproj.contains("excludeMe"))
```
Comment (aciidb0mb3r) [inline Tests/XcodeprojTests/GenerateXcodeprojTests.swift]: This should test the model `Xcode.Project` instead of reading `project.pbxproj`

**Comment by emlai** [comment]
@aciidb0mb3r Thanks for the review! Changed the code as asked. I had to modify the signature of `findDirectoryReferences` a bit. Also removed the `#if os(macOS)`, I realized it wasn't necessary (I think).

File: Sources/Xcodeproj/generate().swift [hunk 3]
```diff
@@ -155,15 +155,17 @@ func open(_ path: AbsolutePath, body: ((String) -> Void) throws -> Void) throws
 }
 
 /// Finds directories that will be added as blue folder
-/// Excludes hidden directories, Xcode projects and reserved directories
-func findDirectoryReferences(path: AbsolutePath) throws -> [AbsolutePath] {
-    let rootDirectories = try walk(path, recursively: false)
+/// Excludes hidden directories, Xcode projects, reserved directories, and
+/// directories specified in "exclude" in the package manifest
+func findDirectoryReferences(forPackage package: Package, fileSystem: FileSystem = localFileSystem) throws -> [AbsolutePath] {
```
Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: Since type is clearly defined we can use `for package` instead of `forPackage package`

Comment (aciidb0mb3r) [inline Sources/Xcodeproj/generate().swift]: So the signature becomes ->
```swift
func findDirectoryReferences(for package: Package, fileSystem: FileSystem = localFileSystem) throws -> [AbsolutePath]
```

**Comment by emlai** [comment]
@aciidb0mb3r Done. 👍

**Comment by aciidb0mb3r** [comment]
Thanks @emlai 
The patch looks good to me. @abertelrud @ddunbar any thoughts here?

I am kind of confused about this since exclude acts as "something you don't want to be picked up the by the convention system" but is still part of the project. For eg Fixtures could be present in a way that they require an entry in excludes but it would make sense and probably will be helpful if they're visible in the Xcode project navigator.

**Comment by ddunbar** [comment]
I'm not 100% confident this is a good idea, for the reason @aciidb0mb3r gave. If it was excluding directories excluded in `.gitignore` then I would agree 100%.

This is a valuable addition, I'm just not sure it is the right default behavior. I'm wondering if we should put it under an option.

**Comment by aciidb0mb3r** [comment]
I agree, lets put this behind an option
@emlai can you make that change?

**Comment by emlai** [comment]
Sure, I can give it a try.

Just to make sure, by "option" you mean an additional command line flag, right?

**Comment by aciidb0mb3r** [comment]
Yup

**Comment by ddunbar** [comment]
Before we just go and do that, let's make sure it is what we want... having a command line option is annoying because users will either always specify it (if they like the behavior for their project) or never specify it for their project (because they don't know about it, or don't like it).

@emlai can you give an example project where you want this behavior? Are the same files in the .gitignore?

**Comment by emlai** [comment]
@ddunbar Yes, in my case the directory is also in .gitignore.

The directory contains files that the program uses to store state between executions of the program.

**Comment by ddunbar** [comment]
In that case, I would rather change this feature to ignore what is in `.gitignore`. We already have a need to build up infrastructure for understanding `.gitignore`, and that will make this something we can just always have on by default.

**Comment by emlai** [comment]
I agree. Having a separate flag for this would be annoying for the reasons you mentioned.

**Comment by abertelrud** [comment]
I definitely agree.  "exclude" means "ignore when building", not that the file has no use from a workflow perspective.  It could be a change log or a readme, for example.  So I agree that doing this via `.gitignore` is the way to go.

**Comment by aciidb0mb3r** [comment]
Closing this PR due to age, feel free to reopen

**Pull request closed without merging**